pkg/provisioning/bg/validation.go

```diff
@@ -0,0 +1,562 @@
+		Name:     "KM Signature Version",
+		function: KMStructureValidSignatureVersion,
+	}
+	kmsigkey = CBnTTest{
+		Name:     "KM Signature Key",
+		function: KMStructureValidSignatureKey,
+	}
+	kmsigscheme = CBnTTest{
+		Name:     "KM Signature Scheme",
+		function: KMStructureValidSignatureSigSchemeCheck,
+	}
+	kmsigkeysize = CBnTTest{
+		Name:     "KM Signature Key Size",
+		function: KMStructureValidSigKeySize,
+	}
+	kmsighashalg = CBnTTest{
+		Name:     "KM Signature Hash Alg",
+		function: KMStructureValidSigHashAlg,
+	}
+	kmsigdata = CBnTTest{
+		Name:     "KM Signature Data",
+		function: KMStructureValidSigData,
+	}
+)
+
+// ImageTests list all CBnTTests in form of a slice
+var ImageTests = []*CBnTTest{
+	&hasFitEntries,
+	&kmValidHeader,
+	&kmValidKMSigOffset,
+	&kmValidReserverd2,
+	&kmrevisionvalid,
+	&kmsvnvalid,
+	&kmidvalid,
+	&kmpubkeyhash,
+	&kmkeyhashcount,
+	&kmbpmkeyhash,
+	&kmsigversion,
+	&kmsigkey,
+	&kmsigscheme,
+	&kmsigkeysize,
+	&kmsighashalg,
+	&kmsigdata,
+}
+
+// HasValidFitEntries defines the behavior for the CBnTTest "HasFitEntries"
+func HasValidFitEntries() (bool, error) {
+	var found int
+	fitentries, err := fit.GetTable(image)
+	if err != nil {
+		return false, err
+	}
+	for _, item := range fitentries {
+		if item.Type() == fit.EntryTypeBootPolicyManifest {
+			found++
+		}
+		if item.Type() == fit.EntryTypeKeyManifestRecord {
+			found += 2
+		}
+	}
+	switch found {
+	case 0:
+		// No KM/BPM
+		return false, fmt.Errorf("No KM and no BPM entry in FIT found")
+	case 1:
+		// NO BPM
+		return false, fmt.Errorf("No BPM entry in FIT found")
+	case 2:
+		// No KM
+		return false, fmt.Errorf("No KM entry in FIT found")
+	case 3:
+		// Success
+		return true, nil
+	default:
+		return false, fmt.Errorf("unexpected Error")
+	}
+}
+
+// KMStructureValidHeader defines the behavior for the CBnTTest "KM Header"
+func KMStructureValidHeader() (bool, error) {
+	var s strings.Builder
+	headerID := "__KEYM__"
+	headerVersion := uint8(0x21)
+	if km.StructInfo.ID.String() != headerID {
+		s.WriteString("KM Header ID incorrect. Have: " + km.StructInfo.ID.String() + " - Want: " + string(headerID) + "\n")
+	}
+	if km.StructInfo.Version != headerVersion {
+		s.WriteString("KM Header Version incorrect. Have: " + string(km.StructInfo.Version) + " - Want: " + string(headerVersion) + "\n")
+	}
+	if km.StructInfo.Variable0 != 0 {
+		s.WriteString("KM Header Variable0 incorrect. Have: " + string(km.StructInfo.Variable0) + " - Want: 0")
+	}
+	if s.Len() != 0 {
+		return false, fmt.Errorf("%s", s.String())
+	}
+	return true, nil
+}
+
+// KMStructureValidKMSOffset defines the behavior for the CBnTTest "KM Signature Offset"
+func KMStructureValidKMSOffset() (bool, error) {
+	var KMSigOffset uint16
+	//Count the bytes to Key Manifest Signature - Everything is static with the exception KeyHash and Key Manifest Signature
+	KMSigOffset = 24
+	// Count KMHashes
+	for _, hash := range km.Hash {
+		// Usage field 8 bytes static
+		KMSigOffset += 8
```

Comment:
can you use binary.Size(some struct) here?

Reply:
When using binary.Size(hash) in the loop, I get -1 (which is not correct obviously) 
Calling binary.Size(km.Hash) also gets me -1. It seems go doesn't consider the structs as fixed length.